README.md
# pmatic

[![PyPI version](https://badge.fury.io/py/pmatic.svg)](https://badge.fury.io/py/pmatic)
[![Build Status](https://travis-ci.org/LarsMichelsen/pmatic.svg?branch=master)](https://travis-ci.org/LarsMichelsen/pmatic)
[![Coverage Status](https://coveralls.io/repos/github/LarsMichelsen/pmatic/badge.svg?branch=master)](https://coveralls.io/github/LarsMichelsen/pmatic?branch=master)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/0b6d7874a5e248a2af685761cccc131c)](https://www.codacy.com/app/lm/pmatic)
[![Code Health](https://landscape.io/github/LarsMichelsen/pmatic/master/landscape.svg?style=flat)](https://landscape.io/github/LarsMichelsen/pmatic/master)

Python API for Homematic. Easy to use.

The [pmatic](https://larsmichelsen.github.io/pmatic/) module provides
access to the Homematic CCU which operates as
the central unit in Homematic based home automation setups. You can use
pmatic directly on the CCU or another system having Python installed.
With pmatic you can write your own Python scripts to communicate with
your CCU device.

## What to do with pmatic?

* Create even small (one line) scripts to read info or trigger actions.
* Execute scripts on any linux system or directly on the CCU.
* Edit scripts in your favorite editor, test on your workstation,
  deploy on another device, for example the CCU, later.
* Code in a very beginner friendly language: Python.
* Organize and control your scripts on the CCU using the pmatic manager

## Why pmatic?

Before I built this API I tried to create a small script to *just* check
all my window sensors, record the time they are opened and then alarm
me to close the window if it was open for too long. No problem I thought.
Lesson learned: It is possible. But only while having a huuuuge pain.
The scripting language is crapy, the web GUI editor misses basic things
like syntax highlighting, undo/redo, auto saving and so on which make
programming comfortable. Last but not least the debugging was a pain
or not possible at all.

Should be possible to make this a lot easier.

I found several other middlewares and libraries for accessing the CCU2
APIs, but most of them required to be executed in somehow specific
environments, were not platform independet or implemented in other crapy
programming languages.

I know sure there is still much room for improvement and cleaner APIs,
but for the moment I think even this small API wrapper is an improvement.

## So how does it work?

pmatic has been implemented in Python. What? Python is not available on
the CCU2, do I need to run it remotely on a separate device now? Yes,
you can. But it is also possible to use it on the CCU2 by installing
a python interpreter with the necessary modules on the device. We'll
get back to it later.

So you have the option to run your pmatic scripts remotely and on the
CCU2. The code stays the same. This means you can develop your scripts
on your workstation, test and debug it using a remote connection to
your CCU2.

You can use all the API methods provided by the CCU2. The data is parsed
and available as python lists or dicts. You can then process the data
in your Python code and use the editor of your choice, use all possible
debugging and profiling features you can imagine with Python.

It's so much fun :-).

Even if you write pmatic in Python, you can also execute custom ReGa
(Homematic Script) through pmatic and also process the output of these
scripts, if you like.

The pmatic manager provides you with a web GUI on your CCU which you can
use to manage (upload, delete, test) your pmatic scripts with. You can
also create scheduling plans in which situations your scripts should be
executed automatically. This can currently be on manager startup, based
on time or based on device events reported by the CCU. Take a look at
the documentation for screenshots and more details.

## Requirements

Pmatic is currently not expecting any special Python modules. pmatic is
supported with Python 2.7, 3.4 and newer. Older versions of Python are
not supported.

There are some which are already use pmatic on Windows systems which have a
Python interpreter installed. But I did not test it and I am pretty sure
there are some changes needed to make it completely work. At least the
pmatic Manager will not work without some changes.

At least the basic functionality of Pmatic has also been tested on OS X
using Python 2.7 and 3.4 from MacPorts. But as for Windows the pmatic Manager
has not been tested on OS X yet.

I am always open to support more platforms. So if one likes to add support
for more, please let me know.

## Installation

Take a look at the [installation documentation](https://larsmichelsen.github.io/pmatic/doc/install.html).

## Documentation

The current [documentation](https://larsmichelsen.github.io/pmatic/doc/index.html)
can be found on the official web site of pmatic.

## Usage

You can find several usage examples in the `examples`. I'll try to add more
in the near future. Some more examples can be found in the
[documentation](https://larsmichelsen.github.io/pmatic/doc/basic_usage.html).

Just a short example:

```
#!/usr/bin/python

import pmatic
ccu = pmatic.CCU()

for device in ccu.devices.query(device_type="HM-Sec-SC"):
    print("%-20s %6s" % (device.name, device.is_open and "open" or "closed"))
```

## What is planned?

Please take a look at the issue tracker and the TODO file.

What really is needed is specific support for the different Homematic devices. I added some
specific classes for devices I have to the `pmatic/devices.py` but have not added properties
and methods to reflect their individual features. And there are also a lot of devices I don't
own. It would be really helpful if you could help out adding more devices to pmatic.

This will make it a lot easier to use pmatic. Because, for example calling `device.is_battery_low`
is a lot more comfortable than digging into the details of a device and find out you have to call
`self.channels[4].values["FAULT_REPORTING"].formated() == "LOWBAT"`.

So please help adding more devices!

## Changes

Please take a look at the [`CHANGELOG.rst`](https://github.com/LarsMichelsen/pmatic/blob/master/CHANGELOG.rst) file for a detailed list of changes.

## Reporting Bugs, Feature Requests

Please use the issue tracker on the [pmatic GitHub page](https://github.com/LarsMichelsen/pmatic).

## Licensing

Copyright © 2016 Lars Michelsen <lm@larsmichelsen.com>. All rights reserved.

All outcome of the project is licensed under the terms of the GNU GPL v2.
Take a look at the LICENSE file for details.


CHANGELOG.rst
The pmatic Changelog
====================

Unreleased (Use snapshot builds to get these changes)
-----------------------------------------------------

General
```````

* FIX: Fixed possble "KeyError: 'device_id'" exeption in set_logic_attributes()
* FIX: Fix incompatibility with current CCU firmware (PMException: Method "room_get_all") #32

Version 0.6 (2017-01-08)
------------------------

General
```````

* Improved low level API error handling when trying to use calls which need arguments
  but using positional arguments instead of named arguments
* CCU package is shipping with ``pytz`` module now
* FIX: Fixed quoting of arguments passed to CCU local low level API
* FIX: Fixed encoding of arguments passed to CCU local low level API

Manager
```````

* Timezone is now configurable. Using this timezone for displaying and time calculations
* FIX: Fixed manual execution of scripts via the "execute scripts" dialog
* FIX: Schedules executed on startup were not shown as triggered and running in GUI
* FIX: Fixed visualization of "keep alive" scripts current state when aborted via GUI
* FIX: Fixed file descriptor inheritance to "external executed" scripts
  (could block manager port during restar)

Version 0.4 (2016-07-14)
------------------------

General
```````

* FIX: Fixed possible endless recursion when listening for device updates
* New helper function for calculating the sun position (``utils.sun_position()``)
  Thanks to Rolf Hempel for implementation!
* New helper function for calculating the dew point (``utils.dew_point()``)
  Thanks to Rolf Hempel for implementation!

Devices
```````

* Added specific device object HM-WDS40-TH-I-2
* Added specific device object HM-Sen-LI-O
* Added specific device object HM-LC-Sw1-Pl-DN-R1
* Added specific device object HM-LC-Bl1PBU-FM

Thanks to Rolf Hempel for adding all of them!

Manager
```````

* Implemented transaction IDs to prevent duplicate execution of actions
  like form submits, script executions or deletion of things
* Improved performance during processing of web pages
* Added yet unfunctional schedule execution based on device events
* Added schedule execution based on device events for devices of a specific type
* FIX: Fixed exception in event processing of non readable values

Version 0.3 (2016-04-19)
------------------------

General
```````

* Added missing install requirements to setup.py: requests, SimpleTR64
* The local/remote detection (``utils.is_ccu()``) is now detecting "local" mode on LXCCU
* Improved error handling of fetched values with the current CCU firmware (2.17.15)
  This firmware fails with error code 501 instead of 601 in case of values that can not
  be fetched for some reason.
* CCU Package: Precompiling all Python files to ``*.pyc`` now for faster initial loading

Devices
```````

* Added specific device object for HM-WDS10-TH-O
* Added specific device object for HM-Sec-SCo
* HM-TC-IT-WM-W-EU: Added missing specific channel ``ChannelWeatherTransmit``

Manager
```````

* Resident and schedule states are now persisted between manager restarts. Please note
  that the current default state directory is ``/var/lib/pmatic`` which is not reboot
  persistant on the CCU. So the resident and schedule states will be reset on reboot.
* Schedule/Resident pages can now only be accessed after setting the manager password.
* Scripts started with "Execute Scripts" can now be run in inline mode
* Schedules: Showing next execution time for timed schedules
* FIX: Fixed custom config via command line not setting new defaults for the argument parsing
  (e.g. ccu_enabled could not be changed to ``False`` using a config file provided by (``-o``).
* FIX: Fixed startup error when connection to CCU is not possible
* FIX: Fixed error during saving of manager config/state files when base directory not
  exists. Trying to create the directory now.
* FIX: Fixed multiple execution of single schedule when multiple timed conditions match
  at the same time
* FIX: Fixed problem in enumeration of schedules/residents after deleting one object

Documentation
`````````````

* Added missing documentation of Parameter classes

Incompatible (possible manual changes needed)
`````````````````````````````````````````````

* Cleanup: Renamed all specific device classes to use underscores
  (e.g. HMPBI4FM to HM_PBI_4_FM)


Version 0.2 (2016-04-07)
------------------------

General
```````

* Added new resident management module ``pmatic.residents`` which can be used to
  manage residents and their attributes. An important feature of this module 
  is to detect the presence of your residents and make your scripts do different
  things depending on the presence of them. Take a look at the docs for details.
* Windows: Made CCU detection platform independent
* Windows: Made setup.py more platform independent
* OS X: Made setup.py and tests work on OS X. Pmatic should be usable on this platform too.
* CCU: Better linking from addon page to the manager page
* FIX: Fixed API call ``room_get_all()`` failing when meta names like "${roomKitchen}" are used.
* FIX: Improved generic error handling for values which are reported to be readable
  but can currently not be read
* FIX: Fixed possible wrong encoding when using ``Pushover.send()``
* FIX: Fixed wrong type for boolean parameter default value

Devices
```````

* HM-CC-RT-DN: Fixed low battery detection via ``device.is_battery_low``
* HM-CC-RT-DN: Added specific attributes/methods:

  * ``device.temperature``
  * ``device.set_temperature``
  * ``device.is_off``
  * ``device.turn_off()``
  * ``device.control_mode``
  * ``device.battery_state``
  * ``device.boost_duration``
  * ``device.set_temperature_comfort()``
  * ``device.set_temperature_lowering()``
  * ``device.valve_state``

* HM-TC-IT-WM-W-EU: Fixed "JSONRPCError: TCL error (601)" when trying to get
  the summary state of this device
* Maintenance channel: Fixed broken ``maintenance_state`` property

Manager
```````


playground.py
#!/usr/bin/env python
# encoding: utf-8
#
# pmatic - Python API for Homematic. Easy to use.
# Copyright (C) 2016 Lars Michelsen <lm@larsmichelsen.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

# http://www.eq-3.de/Downloads/PDFs/Dokumentation_und_Tutorials/HM_XmlRpc_V1_502__2_.pdf
# http://www.eq-3.de/Downloads/Software/HM-CCU2-Firmware_Updates/Tutorials/hm_devices_Endkunden.pdf
# https://sites.google.com/site/homematicplayground/api/json-rpc

# Add Python 3.x behaviour to 2.7
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import sys, pprint
import pmatic
from pmatic.entities import *
from pmatic import utils
from pmatic.exceptions import PMDeviceOffline

#from pmatic.notify import Pushover

# Set log level to debug for pmatic and start logging to console
#pmatic.logging(pmatic.DEBUG)

#ccu = pmatic.CCU(address="http://192.168.1.26", credentials=("Admin", "EPIC-SECRET-PW"))

#print(ccu.signal_strengths)

#for room in ccu.rooms.query(room_name_regex="^Ba"):
#    print(room.name)

sys.exit(1)

#for room in ccu.rooms:
#    print(room.name)
#    for device in room.devices:
#        print(" ", device.name, device.summary_state, device.rssi)
#
#sys.exit(1)

for device in ccu.devices.query(device_name="Zentrale"):
    print(device.__dict__, device.address)

#for device in ccu.devices:
#    print(device.name, device.address, len(device.channels))
#    try:
#        for channel in device.channels:
#            print(" ", channel.name, channel.address, channel.summary_state)
#    except PMDeviceOffline:
#        print("  The device is offline!")

#print(list(ccu.devices))
#print(ccu.devices.addresses())

#ccu.events.init()
#ccu.events.wait()
#ccu.events.close()

sys.exit(1)

# Open up a remote connection via HTTP to the CCU and login as admin. When the connection
# can not be established within 5 seconds it raises an exception.
API = pmatic.api.init(
    address="http://192.168.1.26",
    credentials=("Admin", "EPIC-SECRET-PW"),
    connect_timeout=5,
    log_level=pmatic.DEBUG,
)

#pprint.pprint(API.interface_list_devices(interface="BidCos-RF"))
#sys.exit(1)

# Register with the XML-RPC API of the CCU, then wait until program termination
# for event and print them to the console.
#events = pmatic.events.init(API, listen_address=('192.168.1.36', 9123))
events = pmatic.events.init(API)
events.init()
events.wait()
events.close()

sys.exit(1)

#CCU = CCU(API)
#print(CCU.interfaces())

# Event-Methoden scheinen nicht so zu funktionieren, wie gedacht. Hier kommen jedenfalls keine Events an.
#print(API.event_subscribe())
#
#while True:
#    result = API.event_poll()
#    if result:
#        print(result)
#
#print(API.event_unsubscribe())

#
#for device in Device.get_devices(API, device_name="Büro-Lampe"):
#    for channel in device.channels:
#        print(device.name, channel.get_values())
#        print(device.name, channel.channel_type, channel.name, channel.formated_values())

for device in Device.get_devices(API):
    if not device.is_online:
        print("OFFLINE:", device.name)
    else:
        for channel in device.channels:
            if channel.__class__ == Channel:
                print("", device.name, channel.channel_type, channel.name, channel.summary_state)
                print(channel.values)
            else:
                print(device.name, channel.channel_type, channel.name, channel.summary_state)
                print(channel.values)
                for name, v in channel.values.items():
                    print(name, v.datatype)

sys.exit(1)

import pprint

def Device_listDetailByType(ty):
    device = []
    for device in API.Device_listAllDetail():
        #  {u'address': u'KEQ0163218',
        #  u'channels': [{u'address': u'KEQ0163218:1',
        #                 u'category': u'CATEGORY_SENDER',
        #                 u'channelType': u'SHUTTER_CONTACT',
        #                 u'deviceId': u'1753',
        #                 u'id': u'1774',
        #                 u'index': 1,
        #                 u'isAesAvailable': True,
        #                 u'isEventable': True,
        #                 u'isLogable': True,
        #                 u'isLogged': True,
        #                 u'isReadable': True,
        #                 u'isReady': True,
        #                 u'isUsable': True,
        #                 u'isVirtual': False,
        #                 u'isVisible': True,
        #                 u'isWritable': False,
        #                 u'mode': u'MODE_AES',
        #                 u'name': u'B\xfcro-Fenster',
        #                 u'partnerId': u''}],
        #  u'id': u'1753',
        #  u'interface': u'BidCos-RF',
        #  u'name': u'B\xfcro-Fenster',
        #  u'operateGroupOnly': u'false',
        #  u'type': u'HM-Sec-SC'}
        # pprint.pprint(device)
        if device["type"] == ty:
            devices.append(device)
    return devices

#for device in devices:
#    print device["id"], device["name"]
#    for c in device["channels"]:
#        #{u'category': u'CATEGORY_SENDER', u'index': 1, u'isReady': True, u'isEventable': True, u'name': u'Balkon-Sensor', u'isVirtual': False, u'isWritable': False, u'isVisible': True, u'isLogable': True, u'partnerId': u'', u'deviceId': u'1306', u'isReadable': True, u'address': u'KEQ0174549:1', u'isAesAvailable': False, u'isUsable': True, u'isLogged': True, u'channelType': u'WEATHER', u'id': u'1326', u'mode': u'MODE_DEFAULT'}
#        #print "  ", channel["name"], channel["channelType"], channel["category"], API.Channel_getValue(id=channel["id"])
#        if c["channelType"] == "WEATHER":
#            print "  ", c["name"], "Temp:", API.Channel_getValue(id="BidCos-RF."+c["address"]+".TEMPERATURE"), \
#                  "Humidity:", API.Channel_getValue(id="BidCos-RF."+c["address"]+".HUMIDITY")
#            sys.exit(1)


devices = API.Device_listAllDetail()
import itertools
channels = list(itertools.chain.from_iterable([ d["channels"] for d in devices ]))

def channels_of_room(room_id):
    room = [ r for r in rooms if r["id"] == room_id ][0]
    return [ c for c in channels if c["id"] in room["channelIds"] ]


def avg_of_room(room_id, what):
    val = None
    for c in channels_of_room(room_id):
        if c["channelType"] == "WEATHER":
            this_val = float(API.Channel_getValue(id="BidCos-RF."+c["address"]+"."+what))

        elif c["channelType"] == "CLIMATECONTROL_RT_TRANSCEIVER" and what == "TEMPERATURE":
            this_val = float(API.Channel_getValue(id="BidCos-RF."+c["address"]+".ACTUAL_TEMPERATURE"))

        #elif c["category"] == "CATEGORY_SENDER":
        #    print c["name"], c["channelType"]
        #    continue

        else:
            continue

        if val == None:
            val = this_val
        else:
            val = (val + this_val) / 2
    return val


def avg_rel_humidity_of_room(room_id):
    return avg_of_room(room_id, "HUMIDITY")


def avg_temperature_of_room(room_id):
    return avg_of_room(room_id, "TEMPERATURE")


def outside_rel_humidity():
    outside_room = [ r for r in rooms if r["name"] == "Balkon" ][0]
    return avg_rel_humidity_of_room(outside_room["id"])


def outside_temperature():
    outside_room = [ r for r in rooms if r["name"] == "Balkon" ][0]
    return avg_temperature_of_room(outside_room["id"])


for device in devices:
    if device["type"] == "HM-Sec-SC":
        channel = [ c for c in device["channels"] if c["channelType"] == "SHUTTER_CONTACT" ][0]
        is_open = API.Channel_getValue(id=channel["id"])

        if True: #is_open:
            print((device["name"], repr(channel["id"]), is_open))

            # FIXME: Multiple rooms?
            room = [ r for r in rooms if channel["id"] in r["channelIds"] ][0]
            if not room:
                raise Exception("Found window contact without room")

            room_humidity = avg_rel_humidity_of_room(room["id"])
            room_temperature = avg_temperature_of_room(room["id"])

            if room_humidity:
                print(("Humidity:", room_humidity, "%"))
            if room_temperature:
                print(("Temperature:", room_temperature, "C"))

            if room_humidity and room_humidity > 70:
                pass # lüften!

            if room_temperature and room_temperature < 20:
                pass # fenster zu!

            out_temp = outside_temperature()
            if out_temp > 19:
                pass # wurst, lüfte halt
            elif out_temp > 15:
                pass # 30 minuten
            elif out_temp > 5:
                pass # 15 minuten
            else:
                pass # 10 minuten

            print(("Outside:", outside_rel_humidity(), "%", out_temp, "C"))


API.close()
